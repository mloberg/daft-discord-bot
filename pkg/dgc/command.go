package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
