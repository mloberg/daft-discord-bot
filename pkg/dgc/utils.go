package dgc

import "github.com/bwmarrin/discordgo"

var InteractionResponseEphemeral uint64 = 1 << 6

// GetUser returns the user that created the interaction either in a DM or guild
func GetUser(i *discordgo.InteractionCreate) *discordgo.User {
	if i.Member != nil {
		return i.Member.User
	}

	return i.User
}
