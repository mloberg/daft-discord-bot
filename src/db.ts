import { PrismaClient } from '@prisma/client';

const client = new PrismaClient({
    // TODO: setup logging
});

export default client;
