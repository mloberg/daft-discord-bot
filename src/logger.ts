import fs from 'fs';
import pino from 'pino';

import config from './config';

const logger = pino(
    {
        level: config.logLevel,
        prettyPrint: config.debug,
    },
    config.isTest ? fs.createWriteStream('/dev/null') : process.stderr,
);

export default logger;
