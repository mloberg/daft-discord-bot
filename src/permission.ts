import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
