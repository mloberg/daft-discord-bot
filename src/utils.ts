export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
