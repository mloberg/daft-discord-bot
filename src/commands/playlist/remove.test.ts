import { Client, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import command from './remove';

const mocks = {
    react: jest.fn(),
    logInfo: mocked(logger.info),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Message: jest.fn().mockImplementation(() => ({
        member: { guild: { id: 'test', name: 'Test Guild' } },
        react: mocks.react,
    })),
}));

jest.mock('../../logger');
jest.mock('../../permission');

describe('_playlist delete', () => {
    let message: Message;

    beforeEach(async () => {
        mocks.react.mockClear();
        mocks.logInfo.mockClear();
        mocks.permission.mockClear();

        const client = {} as Client;
        const channel = {} as TextChannel;
        message = new Message(client, {}, channel);

        await db.playlist.create({ data: { name: 'test', guild: 'test', songs: [] } });
        await db.playlist.create({ data: { name: 'foo', guild: 'another', songs: [] } });
    });

    afterEach(async () => {
        await db.$executeRaw`DELETE FROM playlists`;
        await db.$disconnect();
    });

    it('is a command', () => {
        expect(command.name).toBe('playlist delete');
        expect(command).toMatchSnapshot();
    });

    it('deletes a playlist', async () => {
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: ['test'], $0: 'playlist delete' });
        expect(mocks.react).toHaveBeenCalledWith('ðŸŽµ');
        expect(mocks.logInfo).toHaveBeenCalledWith('Deleted playlist test from Test Guild');

        const playlists = await db.playlist.findMany({ where: { name: 'test' } });
        expect(playlists).toHaveLength(0);
    });

    it('throws an error if no name is given', async () => {
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: [], $0: 'playlist remove' })).rejects.toThrow(
            new FriendlyError('Give me a playlist to delete.'),
        );
    });

    it('will not delete a playlist from another guild', async () => {
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: ['foo'], $0: 'playlist remove' });
        const playlists = await db.playlist.findMany({ where: { name: 'foo' } });
        expect(playlists).toHaveLength(1);
    });

    it('will throw an error if user does not have role', async () => {
        mocks.permission.mockReturnValue(false);

        await expect(command.run(message, { _: [], $0: 'playlist delete' })).rejects.toThrow(
            new FriendlyError('You do not have permission to do that.'),
        );
    });
});
