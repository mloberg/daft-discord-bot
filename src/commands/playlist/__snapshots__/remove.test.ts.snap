// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_playlist delete is a command 1`] = `
Object {
  "alias": Array [
    "playlist remove",
  ],
  "description": "Delete a playlist",
  "name": "playlist delete",
  "run": [Function],
  "usage": "<name>",
}
`;
